docs/source/user/configuration.md
# Configuration

## General

Defaults:

```
'general': {
    'jedi': False,
    'register_key_commands': True,
    'page_line_count': 40
}
```


`jedi: bool` - [JEDI](http://jedi.readthedocs.io/en/latest/) is used as a backend for Find usages,
Jump to definition and Show documentation [Scripts](scripts.md). But because JEDI is also used by the Pythonista,
JEDI is not thread safe and I don't know when and how it is used, it's disabled by default. You have to
set it to `True` to enable mentioned scripts.

`register_key_commands: bool` - set to `False` to disable default [Shortcuts](shortcuts.md).

`page_line_count: int` - number of lines to scroll up / down for page up / down.


## Updates

Defaults:

```
'update': {
    'enabled': True,
    'interval': 3600
}
```

`enabled: bool` - set to `False` to disable check for updates.

`interval: int` - check for updates time interval (in seconds).


## File picker

Defaults:

```
'file_picker': {
    'ignore_folders': {
        '': ['.git'],
        '.': ['.Trash', 'Examples',
              'site-packages', 'site-packages-2', 'site-packages-3']
    }
}
```

`ignore_folders: dict` - key is a parent directory (not full path, just name) and value is a list of
folders to ignore. You can use two special values as keys:


* `''` - any parent directory,
* `'.'` - parent directory is `~/Documents`.

Default value says that `.git` folder inside any folder is ignored. `.Trash`,
`Examples`, ... folders inside `~/Documents` folder are ignored as well.

Affects Open quickly and Run quickly scripts, see [Scripts](scripts.md).

## Analyzer

Defaults (since 1.1.0):

* `ignore_codes` option is removed /ignored
* `max_line_length` option is removed / ignored

```
'analyzer': {
    'hud_alert_delay': 1.0,
    'remove_whitespaces': True,
    'flake8': [
        # 1st pass
        ['--select=E901,E999,F821,F822,F823'],
        # 2nd pass
        ['--max-complexity=10', '--max-line-length=127']
    ]
}
```

If `flake8` is provided, `ignore_codes` and `max_line_length` options are ignored. `flake8`
must contain list of passes and every pass contains list of `flake8` arguments. You can run `flake8`
several times with different options in this way.

Defaults (pre 1.1.0):

```
'analyzer': {
    'hud_alert_delay': 1.0,
    'ignore_codes': None,
    'max_line_length': 127,
    'remove_whitespaces': True
}
```

Affects Analyze script, see [Scripts](scripts.md).

## Tester

Defaults:

```
'tester': {
    'hud_alert_delay': 1.0,
    'hide_console': True
}
```

Affects Run unit tests script, see [Scripts](scripts.md).

## Drag and Drop

Defaults:

```
'drag_and_drop': {
    'ignore_folders': {
        '': ['.git'],
        '.': ['.Trash', 'Examples',
              'site-packages', 'site-packages-2', 'site-packages-3', 'stash_extensions']
    }
}
```

Affects Drag & Drop script, see [Scripts](scripts.md).


## Documentation

Defaults:

```
'documentation': {
    'reuse': True,
    'frame': (630, 110, 730, 350)
}
```

`reuse: bool` - same overlay view is reused for consequent show documentation calls if
set to `True`. Otherwise multiple overlays appear in consequent show documentation calls,
but if a symbol's fully qualified name matches existing overlay, this particular overlay is
expanded and activated.

`frame: tuple(float, float, float, float)` - initial overlay frame in the key window coordinates.

Affects Show documentation script, see [Scripts](scripts.md).

## Sample

See [pythonista_startup.py](https://github.com/zrzka/blackmamba/blob/master/pythonista_startup.py)
for more examples.


docs/source/user/configuration.rst
.. _configuration:

#############
Configuration
#############

.. automodule:: blackmamba.config
    :member-order: bysource

CHANGELOG.md
# Change log

## master (unreleased)

* `pythonista_startup.py` sample - execute `main` only if `appex.is_running_extension()` is false

## 1.5.0 (2018-01-11)

* Pass `-p no:cacheprovider` to fix operation not permitted for `.cache` directory (run unit tests script)
* Documentation cleanup and rewrite to Markdown, which is easily editable in Pythonista compared to reST
* `blackmamba.framework.security` introduced
* `blackmamba.uikit.keyboard`
    * Camel case enum constant deprecated, will be removed in 2.0.0
    * Use UPPER_CASED versions, camel case still exists as aliases
* `with bundle()` always unloads `pkg_resources.*` causing issues
    * Launch Pythonista, run Black Mamba Analyze - works vs
    * Launch Pythonista, run Pythonista unit tests, run Black Mamba Analyze - fails, because of `pkg_resources`

## 1.4.3 (2017-12-30)

* Compatibility check with 3.2 (320000)

## 1.4.2 (2017-12-23)

* Compatibility check with 3.1.1 (311017)

## 1.4.1 (2017-11-13)

* Compatibility check with 3.1.1 (311016)
* Bundle `refactoring` enhanced with `future`, `libfuturize`, `libpasteurize` modules
* `script/refactoring/futurize.py` introduced (see [Python Future](https://github.com/PythonCharmers/python-future))
    * Equivalent of `futurize -1 --all-imports -n -w --add-suffix 3 $editor.get_path()` (Stage 1 only)
    * When futurizer ends, editor text is replaced with content of the `.py3` and `.py3` is trashed
    * You can run futurizer script with `Cmd Option F`
* Improved updates check
    * Console is not cluttered with local / latest release info (installer prints this)
    * Update `check()` doesn't ask if update should (not) be installer (installer also asks)
    * If there's new update available, installer is executed, you will still be asked (just once, not twice)
* Script `new_file.py` modified
    * File opened
        * Asks for a file name (empty & `Enter` -> Cancel)
        * New file path is currently opened file dirname + entered file name
        * If file doesn't exist, new file is created and opened
        * If file exists, file is opened
    * No file opened
        * Same behavior as now
        * New tab created and _New File..._ button tap emulated
* Installer
    * Replaced `ModuleNotFoundError` (Python 3.6) with `ImportError` (Python 3.5)

## 1.4.0 (2017-11-01)

* Fixed keyboard shortcut selector name generator
* Bundle `refactoring` introduced (includes `rope`)
* Refactoring functions introduced
    * It's an EXPERIMENT. You should use version control system to avoid loosing data.
    * `Cmd Option O` - Organize imports
    * `Cmd Option E` - Expand star imports
    * `Cmd Option R` - Rename identifier
    * Can be used as scripts as well, see `script/refactoring` folder
    * Preview dialog can be closed with `Cmd .` / `Esc`, confirmed with `Enter`
* Scroll back to initial cursor location if analyzer didn't find an issue, ugly, but better than end of the file

## 1.3.3 (2017-10-27)

* Installer improved
    * Print installed version to the console
    * Show installed version in the dialog update/replace dialog
    * Dialog content differs for update/replace (version comparison)
* `script/selfupdate.py` added (it just runs installer via requests)
* Check for updates
    * Update dialog allows you to run `selfupdate.py`
    * If you hit cancel, installer command is copied to the clipboard
      and additional info about `selfupdate.py` / installer command
      is printed to the console

## 1.3.2 (2017-10-25)

* Fixed exception when there's empty tab opened (#30)
* Open quickly selects proper tab when the file is already opened
  and empty tab exists as well
* Compatibility check with 311015
    * `Cmd-Shift-]`, `Cmd-Shift-[` registered only in Pythonista < 311015
    * These shortcuts are natively supported
* flake8-3.5.0, pyflakes-1.6.0
    * Bare exception check
    * Ambigious identifier check  

## 1.3.1 (2017-10-19)

* Run scripts via `pythonista3://`, not just `pythonista://` (see #29)

## 1.3.0 (2017-10-19)

* Pythonista 3.1.1 (311014)
    * Compatibility check with 311014
    * Shortcuts `Cmd W`, `Ctrl [Shift] Tab` work
* Trashed `Cmd Q` shortcut (workaround for 311013)
* Open Quickly supports File System Provider bookmarks
    * You can search / open files added to Pythonista via `Open...` (External files)
    * These files are added as bookmarks (iOS terminology)
    * If cell subtitle starts with `Documents`, it's a file from `~/Documents`
    * If cell subtitle starts with `Bookmark`, it's a file from FSP

## 1.2.2 (2017-10-18)

* Bundled `lib/pep8` removed, unused
* Fixed `get_actions` (exception when user has no custom actions)
* Pythonista 3.1.1 (311013)
    * Compatibility check with 311013
    * Shortcuts `Cmd W`, `Ctrl Tab`, `Ctrl Shift Tab` no longer work (1)
    * Shortcuts `Cmd 1..9`, `Cmd Shift ]`, `Cmd Shift [`, `Cmd Shift W` still work

(1) Pythonista 311013 provides these shortcuts natively. Unfortunately,
they do not work. Also these shortcuts are provided elsewhere in the
responder chain, so, even if I register them via the Black Mamba, responder
chain catches them sooner then Black Mamba and they do not work.

All these shortcuts do work prior to 311013. If you have 311013 installed, you
can use `Cmd Q` to close tab (temporary) and `Cmd Shift ]` / `Cmd Shift [`
to show next / previous tab.

## 1.2.1 (2017-10-11)

* [Bundled packages](http://blackmamba.readthedocs.io/en/stable/bundled.html) links and licenses
* `ide.run_script` respects `delay` argument (did contain hardcoded value)
* Analyze script documentation mentions `flake8`
* Drag & Drop script
    * Children nodes are lazy-loaded (faster)
    * Dropped folder reloads ...
        * Target row if not expanded (to display triangle)
        * Children nodes (dropped folder appears if it didn't exist before drop)
* Show documentation
    * Do not show picker / docstring if symbol was found, but it has no docstring


## 1.2.0 (2017-10-10)


blackmamba/config.py
#!python3
from collections import Mapping
from copy import deepcopy

__all__ = ['get_config_value']

_DEFAULTS = {
    'general': {
        'jedi': False,
        'register_key_commands': True,
        'page_line_count': 40
    },
    'update': {
        'enabled': True,
        'interval': 3600
    },
    'file_picker': {
        'ignore_folders': {
            '': ['.git'],
            '.': ['.Trash', 'Examples',
                  'site-packages', 'site-packages-2', 'site-packages-3']
        }
    },
    'analyzer': {
        'hud_alert_delay': 1.0,
        'remove_whitespaces': True,
        'flake8': [
            # 1st pass
            ['--select=E901,E999,F821,F822,F823'],
            # 2nd pass
            ['--max-complexity=10', '--max-line-length=127']
        ]
    },
    'tester': {
        'hud_alert_delay': 1.0,
        'hide_console': True
    },
    'drag_and_drop': {
        'ignore_folders': {
            '': ['.git'],
            '.': ['.Trash', 'Examples',
                  'site-packages', 'site-packages-2', 'site-packages-3', 'stash_extensions']
        }
    },
    'documentation': {
        'reuse': True,
        'frame': (630, 110, 730, 350)
    }
}

_config = dict(_DEFAULTS)


def get_config_value(key_path, default=None):
    v = _config
    try:
        for k in key_path.split('.'):
            v = v[k]
        return v
    except TypeError:
        return default
    except KeyError:
        return default


def _update(d, u):
    for k, v in u.items():
        if isinstance(d, Mapping):
            if isinstance(v, Mapping):
                d[k] = _update(d.get(k, {}), v)
            else:
                d[k] = v
        else:
            d = {k: v}
    return d


def update_config_with_dict(d):
    global _config
    _config = _update(deepcopy(_config), d)
